src/org/omegat/core/machinetranslators/ApertiumTranslate.java

```diff
@@ -138,4 +161,79 @@ protected String getJsonResults(String json) {
 
         return tr;
     }
+
+    /**
+     * Whether or not to use a custom Apertium server
+     */
+    private boolean useCustomServer() {
+        String value = System.getProperty(PROPERTY_APERTIUM_SERVER_CUSTOM,
+                Preferences.getPreference(PROPERTY_APERTIUM_SERVER_CUSTOM));
+        return Boolean.parseBoolean(value);
+    }
+
+    /**
+     * Get the custom server URL
+     */
+    private String getCustomServerUrl() {
+        String value = System.getProperty(PROPERTY_APERTIUM_SERVER_URL,
+                Preferences.getPreference(PROPERTY_APERTIUM_SERVER_URL));
+        return value;
+    }
+
+    @Override
+    public boolean isConfigurable() {
+        return true;
+    }
+
+    @Override
+    public void showConfigurationUI(Window parent) {
+
+        JCheckBox apiCheckBox = new JCheckBox(OStrings.getString("APERTIUM_CUSTOM_SERVER_LABEL"));
+        apiCheckBox.setSelected(useCustomServer());
+
+        MTConfigDialog dialog = new MTConfigDialog(parent, getName()) {
+            @Override
+            protected void onConfirm() {
+                boolean temporary = panel.temporaryCheckBox.isSelected();
+                System.setProperty(PROPERTY_APERTIUM_SERVER_CUSTOM, Boolean.toString(apiCheckBox.isSelected()));
+                Preferences.setPreference(PROPERTY_APERTIUM_SERVER_CUSTOM, apiCheckBox.isSelected());
+                String server = panel.valueField1.getText().trim();
+                String apiKey = panel.valueField2.getText().trim();
+
+                if (!getCustomServerUrl().equals(server)) {
+                    clearCache();
+                }
+
+                System.setProperty(PROPERTY_APERTIUM_SERVER_URL, server);
+                Preferences.setPreference(PROPERTY_APERTIUM_SERVER_URL, server);
+                setCredential(PROPERTY_APERTIUM_SERVER_KEY, apiKey, temporary);
```

Comment:
With this code we can have a situation where `PROPERTY_APERTIUM_SERVER_CUSTOM` is `true` but there is no value for `PROPERTY_APERTIUM_SERVER_URL` or `PROPERTY_APERTIUM_SERVER_KEY`.

There are a lot of different levels of hand-holding you can do to solve this. Right now I guess this is handled by the code path that shows `APERTIUM_CUSTOM_SERVER_NOTFOUND`. If that's good enough then you can leave it.

If you wanted to be nicer, I would add `DocumentListener`s to the text fields that enable/disable the OK button according to whether the required fields are filled in:

```java
Runnable updateOk = () -> {
    boolean needsFields = apiCheckBox.isSelected();
    boolean hasFields = !dialog.panel.valueField1.getText().trim().isEmpty()
            && !dialog.panel.valueField2.getText().trim().isEmpty();
    boolean canConfirm = !needsFields || hasFields;
    dialog.panel.okButton.setEnabled(canConfirm);
};
// Call updateOk.run() in listeners
```

It should probably also be called in the `ItemListener` below.

Reply:
Thanks, I have added this check to prevent users from clicking the OK button with a blank URL field (the API key is optional, so it is not checked). If the URL is not valid the ```APERTIUM_CUSTOM_SERVER_NOTFOUND``` message is shown to prompt the user to check it.

Reply:
If the API key is not supplied then the URL will have `&key=` with a blank value. I take it that's not a problem?

Reply:
Yes, it is fine, the API itself handles this nicely.